toolchain/check/handle_function.cpp

```diff
@@ -55,13 +56,110 @@ static auto BuildFunctionDecl(Context& context, bool is_definition)
       context.node_stack().PopIf<Parse::NodeKind::ImplicitParamList>().value_or(
           SemIR::InstBlockId::Empty);
   auto name_context = context.decl_name_stack().FinishName();
-  auto fn_node =
-      context.node_stack()
-          .PopForSoloParseNode<Parse::NodeKind::FunctionIntroducer>();
+  context.node_stack()
+      .PopAndDiscardSoloParseNode<Parse::NodeKind::FunctionIntroducer>();
+
+  // Process modifiers and introducer.
+  auto first_node = context.innermost_decl().first_node;
+  auto modifiers = ModifiersAllowedOnDecl(context,
+                                          KeywordModifierSet()
+                                              .SetPrivate()
+                                              .SetProtected()
+                                              .SetAbstract()
+                                              .SetDefault()
+                                              .SetFinal()
+                                              .SetImpl()
+                                              .SetVirtual(),
+                                          "`fn` declaration");
+  switch (context.containing_decl().kind) {
+    case DeclState::FileScope:
+      modifiers =
+          ModifiersAllowedOnDecl(context, KeywordModifierSet().SetPrivate(),
+                                 "`fn` declaration at file scope");
```

Comment:
In general, it looks like all these diagnostics start with "`fn` declaration". Structurally, that's true in general. Perhaps the keyword lex token kind could be passed, and only the "at file scope" part would need to be provided? Maybe that latter part could be structural, too, if there's duplication across declarations?

Part of this is making it easier to process diagnostics programmatically, without string parsing.